src/main/java/uk/gov/justice/digital/hmpps/keyworker/rolemigration/RoleAssignmentsService.java

```diff
@@ -0,0 +1,77 @@
+package uk.gov.justice.digital.hmpps.keyworker.rolemigration;
+
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Service;
+import org.springframework.validation.annotation.Validated;
+import org.springframework.web.client.HttpClientErrorException;
+import uk.gov.justice.digital.hmpps.keyworker.dto.RoleAssignmentsSpecification;
+
+import javax.validation.Valid;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+@Service
+@Slf4j
+@Validated
+public class RoleAssignmentsService {
+
+    private final RoleService roleService;
+
+    public RoleAssignmentsService(RoleService roleService) {
+        this.roleService = roleService;
+    }
+
+    /**
+     * Look up users by caseload (prison) and membership of specification.rolesToMatch.
+     * for each matched user:
+     * Assign to the user the roles identified in specification.rolesToAssign.
+     * Remove from the user (at the caseload) the roles identified by specification.rolesToRemove.
+     */
+    @PreAuthorize("hasAnyRole('MAINTAIN_ACCESS_ROLES', 'MAINTAIN_ACESS_ROLES_ADMIN')")
+    public void updateRoleAssignments(@Valid final RoleAssignmentsSpecification specification) {
+        log.debug("Updating role assignments: {}", specification);
+        for (val caseload : specification.getCaseloads()) {
+
+            val usernamesForCaseload = findUsernamesMatchingRolesAtCaseload(specification.getRolesToMatch(), caseload);
+            log.debug("Found {} users for the {} caseload: {}.", usernamesForCaseload.size(), caseload, usernamesForCaseload);
+            usernamesForCaseload.forEach(username -> {
+                try {
+                    assignRolesToUser(username, specification.getRolesToAssign());
```

Comment:
what will happen if the user already has the role - i.e. the code is re-run?

Reply:
Its also worth considering the restartability of this API.  ie. if one nomis call fails, do we stop, retry, etc.

Reply:
The end-point works by caseload, then by matched user within caseload.
The end-point calls elite2api end-points to add and remove roles from users (at caseloads).
The remote calls are repeatable; you can add the same role as many times as you like. Removing a role from a user who does not have the role results in a 404.
The end-point swallows the 404 errors and continues. 
More serious errors stop processing for the current user at the current caseload. Work resumes with the next user/caseload combination.  Adds are done before removes - so all roles should be added to a user before any roles are removed from that user.
Therefore the end-point can be used repeatedly with the same payload, and eventually, (on first call if the remote calls are all OK) the state of the role assignments should settle to what is specified in the payload.
I've tried various combinations of add/remove in different sequences against T3.  It seems to work reasonably well.
The logs should give plenty of information about what happened.
I can demo on Tuesday if that helps...